### ISSUE TITLE
Security of worker_threads + vm.Module + deleting all globals?

### ISSUE THREAD

**Issue by idleman** [nodejs/help#1378]
Running untrusted code using only the vm module is not safe, but what about if it is used together with worker_threads and all global variables are deleted before the untrusted code is evaluated? See below for an example:

```
// main.mjs 
import worker_threads from 'worker_threads';
import path from 'path';

const { Worker, isMainThread } = worker_threads;

if(isMainThread) {
  const worker = new Worker(path.resolve('./main.mjs'));
} else {
  const globalScope = (new Function('return this'))();
  const global = {};
  Object.keys(globalScope).forEach(name => {
    global[name] = globalScope[name];
    delete globalScope[name]
  });
  import('./untrusted')
    .then(exports => exports.default(global))
    .catch(console.error.bind(console))
}


// untrusted.mjs:
import vm from 'vm';

export default async function container(global) {

  const code = `
    // this code can be anything, how secure is it safe?
    const globalScope = (new Function('return this')());
    console.log('Global variables:', Object.keys(globalScope));  
    const process = globalScope.constructor.constructor('return Object.keys(this)')();
    try {
      // will not work
      process.exit();
    } catch(err) { }
    
    for(;;) ; // infinite loops will not work either
  `;

  const context = vm.createContext({
    console: {
      log: (...args) => {
        console.log(...args);
        return;
      }
    }
   });


  const module = new vm.Module(code, { context });
  await module.link(() => null);
  module.instantiate();
  try {
    await module.evaluate({ timeout: 50 });
  } catch(err) {
    console.error(err); // will catch the timeout error
  }
};

```
Is it safe?

**Comment by devsnek** [comment]
starting with the basics:
1) worker_threads isn't a security mechanism
2) vm isn't a security mechanism

based on these two items alone i can say that there is no guarantee that someone will be unable to "break out" of what you consider the sandbox no matter what you do using those two mechanisms.

beyond that there are so many ways you can mess with node core... you could define accessors in unexpected places on global objects you got via literals or other things and chain your way up to whatever you want. ways you and i haven't thought of to do weird random stuff that adds up to your system getting pwned.

if you want to be truly secure, create a native module that lets your create your own v8::Isolate

**Comment by idleman** [comment]
I am deleting every global object the code can access and each Worker has its own v8::Isolate.

Can you provide some examples on how you can "break out" in the example I posted, by only changing the "code" variable?

**Comment by devsnek** [comment]
> each Worker has its own v8::Isolate.

yes but node then adds back in all the tools that let you use the net and fs and such.

> Can you provide some examples on how you can "break out" in the example I posted, by only changing the "code" variable?

like i said above you could mess with global properties in strange ways that node core doesn't expect, or that host code doesn't expect. i don't really feel like digging around to find insecurities in node core or your code, but i can guarantee you they exist and that people will find ways to exploit them. don't ever assume you're smarter than your attackers.

your code is probably fine in the common case but as i said, you aren't using any security mechanisms anywhere so don't assume its bullet-proof. Intel thought their context switching was secure and [look how that turned out](https://meltdownattack.com/).

**Comment by idleman** [comment]
So basically do it comes down if malicious code can be embedded in ECMAScript global objects such as Map/Set/ArrayBuffer that nodejs core uses internally to trick it to bad stuff more than just crash. Is this correct?

@trvsapjiac

> Also, what is the purpose of the global argument that is passed to the exported async function container? You are not using it in the function, you probably want to delete properties after the object becomes contextified. And note that {} is still not an empty object, but Object.create(null) is.

I have forgot too remove it. But the plan was to provide it if the function would need it.

**Comment by trvsapjiac** [comment]
> So basically do it comes down if malicious code can be embedded in ECMAScript global objects such as Map/Set/ArrayBuffer that nodejs core uses internally

I think what @͏devsnek said is self-explanatory. The proof that it is safe must cover all possible cases, which (given how Node.js works and how JavaScript is designed) is very hard. When talking about "arbitrary code", we usually think about code that we use in real applications. But, JavaScript specification is far beyond it. It has been proven that JavaScript [can read](https://www.youtube.com/watch?v=I5mRwzVvFGE) physical ram memory, so I don't see the point of talking about safety, even in the JavaScript.

There are a lot of security mechanisms v8 engine performs in order to sandbox a script. Scripts are parsed and converted into bytecode and executed in the environment associated with an isolate. All isolates that Node.js uses contain a permanent reference to the `process.binding`. Therefore, it is theoretically possible to obtain reference to it from any script executed by Node.js, both from cjs module, vm context or worker thread.

> Can you provide some examples on how you can "break out"

It's not easy to find an example, but it definitely doesn't mean there are no examples. But consider this untrusted code:

```js
var a = [];
while(1)
  a.push(String(a.length).repeat(1e8));
```

It will crash your process no matter whether it is executed in the main thread, worker thread or vm.

On the other hand, the `console` is also some kind of a wrapper around `process.stdout`. When you log something, Node.js firstly calls `util.inspect` to stringify it (except in special cases) and then transforms it into a buffer and passes to the stdout. One may do a lot of hacks there, from overriding Buffer prototype methods, typed array methods, defining custon inspect function, to injecting a Proxy instance into the getter/setter, etc. Many of such attempts would probably not work, but it's not hard to believe that some of them may bring you reference to the process bindings.

In conclusion, [creating a new isolate](https://v8docs.nodesource.com/node-9.3/d5/dda/classv8_1_1_isolate.html) and executing untrusted code inside it is probably the best option you can choose.

**Comment by addaleax** [comment]
Btw, maybe to expand on what was already said:

- Threading + vm contexts is insufficient isolation, but it’s also not quite the worst thing. 
- `console` would *not* point to Node’s `console`, but the V8 built-in that only reports to the inspector – this should be okay.
- `require` is not available in vm contexts, so that’s not an issue.
- Out-of-memory situations are problematic. This could be addressed in the future, but at this point, V8 requires the process to crash if this happens.
- Spectre/meltdown situations are problematic. You can *build* Node.js with specific options that mitigate the impact (`v8_untrusted_code_mitigations` in `config.gypi`), but by default, those are disabled.

I’d like to see workers + vm eventually become a good isolation environment, but we’re definitely not quite there yet.

> In conclusion, creating a new isolate and executing untrusted code inside it is probably the best option you can choose.

It’s no better or worse than Workers + vm.

**Comment by idleman** [comment]
I understand, thanks for the help everyone!
